File: 프로그래머스/lv3/49191. 순위/순위.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int n, vector<vector<int>> results) {
    int answer = 0;
    vector<vector<int>> board(n + 1, vector<int>(n + 1, 0));
    
    for(vector<int> result: results) {
        board[result[0]][result[1]] = 1;
        board[result[1]][result[0]] = -1;
    }
    
    for(int k = 1; k <= n; k++) {
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= n; j++) {
                if(board[i][j] == 0) {
                    if(board[i][k] == 1 && board[k][j] == 1) {
                        board[i][j] = 1;
                    }
                    else if(board[i][k] == -1 && board[k][j] == -1) {
                        board[i][j] = -1;
                    }
                }
                
            }
        }
    }
    
    for(int i = 1; i <= n; i++) {
        int cnt = 0;
        for(int j = 1; j <= n; j++) {
            if(board[i][j] != 0) cnt++;
        }
        if(cnt == n - 1) answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/49191. 순위/순위.py
def solution(n, results):
    graph = [[0 for _ in range(n)] for _ in range(n)]
    
    for A, B in results:
        graph[A - 1][B - 1] = 1
        graph[B - 1][A - 1] = -1
        
    for i in range(n):
        for x in range(n):
            for y in range(n):
                if graph[x][y] == 0:
                    if graph[x][i] == 1 and graph[i][y] == 1:
                        graph[x][y] = 1
                    if graph[x][i] == -1 and graph[i][y] == -1:
                        graph[x][y] = -1
                        
    answer = sum([
            1 if sum([1 if result != 0 else 0 for result in row]) == n - 1
            else 0
            for row
            in graph
        ])
    return answer